crates/c-api/src/wasi.rs

```diff
@@ -149,6 +154,121 @@ pub extern "C" fn wasi_config_inherit_stdout(config: &mut wasi_config_t) {
     config.builder.inherit_stdout();
 }
 
+struct CustomOutputStreamInner {
+    foreign_data: crate::ForeignData,
+    callback: extern "C" fn(*mut c_void, *const u8, usize) -> isize,
+}
+
+impl CustomOutputStreamInner {
+    pub fn write(&self, buf: &[u8]) -> isize {
+        (self.callback)(self.foreign_data.data, buf.as_ptr(), buf.len())
+    }
+}
+
+pub struct CustomOutputStream {
+    inner: std::sync::Arc<CustomOutputStreamInner>,
+}
+
+impl CustomOutputStream {
+    pub fn new(
+        foreign_data: crate::ForeignData,
+        callback: extern "C" fn(*mut c_void, *const u8, usize) -> isize,
+    ) -> Self {
+        Self {
+            inner: std::sync::Arc::new(CustomOutputStreamInner {
+                foreign_data,
+                callback,
+            }),
+        }
+    }
+}
+
+impl Clone for CustomOutputStream {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl wasmtime_wasi::p2::Pollable for CustomOutputStream {
+    async fn ready(&mut self) {}
+}
+
+#[async_trait::async_trait]
+impl wasmtime_wasi::p2::OutputStream for CustomOutputStream {
+    fn write(&mut self, bytes: Bytes) -> Result<(), StreamError> {
+        let wrote = self.inner.write(&bytes);
+
+        if wrote < 0 {
+            return Err(StreamError::Trap(
```

Comment:
This'll want to use `LastOperationFailed` rather than `Trap` to match the behavior of `AsyncWrite` below